File: killrvideo-service-ratings/src/main/java/com/killrvideo/service/rating/dao/VideoRatingDao.java
package com.killrvideo.service.rating.dao;

import com.datastax.oss.driver.api.mapper.annotations.Dao;
import com.datastax.oss.driver.api.mapper.annotations.Increment;
import com.datastax.oss.driver.api.mapper.annotations.Select;
import com.datastax.oss.driver.api.mapper.annotations.StatementAttributes;
import com.killrvideo.service.rating.dto.VideoRating;

import java.util.Optional;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;

@Dao
public interface VideoRatingDao {
    @Select
    @StatementAttributes(consistencyLevel = "LOCAL_ONE")
    CompletableFuture<Optional<VideoRating>> findRating(UUID videoId);

    @Increment(entityClass = VideoRating.class)
    @StatementAttributes(consistencyLevel = "LOCAL_QUORUM")
    CompletableFuture<Void> increment(UUID videoid, long ratingCounter, long ratingTotal);
}

File: killrvideo-service-ratings/src/main/kotlin/com/killrvideo/service/rating/dao/VideoRatingDao.kt
package com.killrvideo.service.rating.dao

import com.datastax.oss.driver.api.mapper.annotations.Dao
import com.datastax.oss.driver.api.mapper.annotations.Increment
import com.datastax.oss.driver.api.mapper.annotations.Select
import com.datastax.oss.driver.api.mapper.annotations.StatementAttributes
import com.killrvideo.service.rating.dto.VideoRating
import java.util.*
import java.util.concurrent.CompletableFuture

@Dao
interface VideoRatingDao {
    @Select
    @StatementAttributes(consistencyLevel = "LOCAL_ONE")
    fun findRating(videoId: UUID): CompletableFuture<VideoRating?>

    @Increment(entityClass = [VideoRating::class])
    @StatementAttributes(consistencyLevel = "LOCAL_QUORUM")
    fun increment(videoid: UUID, ratingCounter: Long, ratingTotal: Long): CompletableFuture<Void>
}
